atomic_reactor/plugins/post_compare_components.py

```diff
@@ -0,0 +1,81 @@
+"""
+Copyright (c) 2017 Red Hat, Inc
+All rights reserved.
+
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
+from atomic_reactor.plugin import PostBuildPlugin, PluginFailedException
+from atomic_reactor.constants import (PLUGIN_COMPARE_COMPONENTS_KEY,
+                                      PLUGIN_FETCH_WORKER_METADATA_KEY)
+
+
+class CompareComponentsPlugin(PostBuildPlugin):
+    """
+    Compare components from each worker build and verify the same version was
+    on each worker.
+    """
+
+    key = PLUGIN_COMPARE_COMPONENTS_KEY
+    is_allowed_to_fail = False
+
+    def rpm_compare(self, a, b):
+        """
+        Compare rpm component version 'a' with 'b'.
+        'name' is implied equal as it is the key() lookup
+        """
+        if a['version'] == b['version'] and \
+           a['release'] == b['release'] and \
+           a['signature'] == b['signature']:
+            return
+
+        raise PluginFailedException("Version mismatch for package %s" % a['name'])
+
+    def get_buildroot(self, worker_metadatas):
+        """
+        Find the buildroot entry of the metadata.
+
+        :return: list, containing dicts of partial metadata
+        """
+        buildroots = []
+        for platform in sorted(worker_metadatas.keys()):
+            for instance in worker_metadatas[platform]['buildroots']:
+                buildroots.append(instance)
+
+        return buildroots
+
+    def run(self):
+        """
+        Run the plugin.
+        """
+
+        worker_metadatas = self.workflow.postbuild_results.get(PLUGIN_FETCH_WORKER_METADATA_KEY)
+        buildroots = self.get_buildroot(worker_metadatas)
+        self.log.debug("DON: md is: %s and br are %s" % (worker_metadatas, buildroots))
+
+        # master compare list
+        master_comp = {}
+
+        # The basic strategy is to start with empty lists and add new component
```

Comment:
+1 for this algorithm

Reply:
I think `master_comp` keys ought to be tuples of (type,  name).

Reply:
Ah! So it can be used like this?
```python
identifier = (t, name)
if identifier not in master_comp:
  master_comp[identifier] = component
  continue
```

Type wasn't compared before, but since we're comparing it now, I think tuples as keys make a lot of sense.

Reply:
Yes, exactly like that.

Reply:
Hmm, bad 'types' don't fail like they are supposed to.  And maybe we don't care in some scenarios.

idA = (foo, perl)
idB = (rpm, perl)

should this fail?  We can pre-filter certain 'type' before saving, ie

if t not in accepted_types:
  raise(Bad type)

But
idA = (log, perl)
idB = (rpm, perl)

would pass because 'log' would be acceptable type or replace 'log' with 'jar'??  Are we ok with that limitation?

Cheers,
Don

Reply:
The idea is that the unique identifier for a component is *both* the type and the name. In your examples above, `idA` and `idB` would never be compared to each other. However, both can still be stored in `master_comp` for later usage when another component with same type and name is found.

Reply:
Sure, I get that.  My point was (and I am probably going overboard with this scenario), if/could a component come in with a bad component type?  If so, that gets missed because it would be assumed to be a different component.  Now if we accept that atomic_reactor provides the component type correctly on the worker (and a corrupt json file doesn't happen in between), then we have nothing to worry about.

I just wanted to throw that out there.  I believe my latest code implements the scheme you are requesting with the addition of a sanity 'type' check right above it.

Cheers,
Don

Reply:
We need to trust that the data is accurate. What we're trying to look for is an RPM component that has unexpectedly been installed with a different n-v-r.

Reply:
I guess bad experience has taught to me careful with pulling data over the network.  In retrospect, I wish we implemented a simple checksum scheme for the metadata to verify it with fetch_worker_metadata.  That would make my argument moot.

Regardless of my opinion, I believe current code implements what you requested?  Or is there something else I need to change?

Cheers,
Don